[Performance] Refactor task list state management and visibility checks to reduce unnecessary tasklist API call

chihsuan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #55127 

The task list API endpoint `/index.php/wp-json/wc-admin/onboarding/tasks?_locale=user` is currently called on every WooCommerce admin page load. While this is necessary when task lists are visible and incomplete (due to the Activity Panel), these API calls are redundant when task lists are either hidden or completed.

This PR optimizes performance by eliminating unnecessary API calls through smarter state management and visibility checks.

#### 1. New Hook: `useTaskListsState` for Activity panel
Introduced a centralized hook to manage task list state and visibility:
- Returns cached default state for hidden/completed task lists
- Only triggers API calls for visible and incomplete task lists

#### 2. New Utility Functions
Added helper functions to improve code clarity and reusability:
- `isTaskListVisible(taskListId: string)`: Checks if a task list should be shown
- `isTaskListCompleted(taskListId: string)`: Verifies completion status
- `isTaskListActive(taskListId: string)`: Combined visibility/completion check

#### 3. Technical Cleanup
- Removed deprecated `useActiveSetupList` hook previously used for task list experiments
- Consolidated task list visibility logic into a single source of truth


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:


1. Create a fresh site with this branch
2. Skip core profiler
3. Go to any WooCommerce admin page and check network requests
   - Verify task list API is called when lists are visible
   - Verify the `Finish setup` tab is shown on the activity panel and the tasklist is visible when clicking on the tab
4. Hide both setup and extended task lists
   - Verify task list API is NOT called after hiding
   - Verify the `Finish setup` tab is not shown on the activity panel
5. Re-enable and complete the setup and extended task lists
   - Verify task list API is NOT called on WC admin pages (except for homescreen) after completion
   - Verify the `Finish setup` tab is NOT shown on the activity panel
   - Verify the Homescreen layout is two-column after completing the task lists
<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

chihsuan committed:
Refactor task list state management and visibility checks

This commit introduces a new `useTaskListsState` hook to centralize task list state management and simplify visibility checks across multiple components. Key changes include:

- Create a new `use-tasklists-state.js` hook to consolidate task list state logic
- Replace direct data store queries with the new hook in various components
- Use `visibleTaskListIds` setting to determine task list visibility
- Improve state handling for setup and extended task lists
- Remove deprecated `useActiveSetupTasklist` hook

chihsuan committed:
Add changelog

chihsuan committed:
Fix lint

chihsuan committed:
Merge branch 'trunk' into perf/tasklist-api

chihsuan committed:
Use ts and add tests

chihsuan committed:
Refactor task list state management and visibility checks

chihsuan committed:
Add completed task list IDs to preloaded settings

This change updates the task list preloaded settings to include both visible and completed task list IDs, improving the frontend's ability to track task list state.

chihsuan committed:
Add function to count things to do next in task list

Add a utility function `getThingsToDoNextCount` to calculate the number of remaining tasks in an extended task list, improving task management functionality.

chihsuan committed:
Fix tests

chihsuan committed:
Refactor homescreen layout and activity panel task list handling

- Improve task screen detection logic in layout component
- Remove unnecessary import and use existing task list state hook
- Update activity panel to use task list completion check

chihsuan committed:
Refactor task list visibility and rendering logic

- Update task list state checks to use new `isTaskListCompletedOrHidden` function
- Modify header and deprecated tasks components to use new visibility logic
- Simplify deprecated tasks rendering by removing unnecessary state management
- Add early return for deprecated tasks plugin when task lists are completed or hidden

chihsuan committed:
Merge branch 'trunk' into perf/tasklist-api

chihsuan committed:
Update task list visibility checks and imports

- Replace `isTaskListCompletedOrHidden` with `isTaskListVisible` in activity panel
- Simplify conditional rendering in homescreen layout using task list state
- Remove unnecessary import in display options file

chihsuan commented on plugins/woocommerce/src/Admin/Features/OnboardingTasks/TaskList.php:
```diff
@@ -168,10 +168,23 @@ public function is_hidden() {
+		if ( ! $this->visible ) {
 			return false;
 		}
-		return ! $this->is_hidden();
```
The last `$this->is_hidden()` is redundant. Remove it to improve performance and make the code cleaner.

chihsuan commented on plugins/woocommerce/src/Admin/Features/OnboardingTasks/TaskList.php:
```diff
@@ -341,7 +354,7 @@ public function possibly_track_completion() {
 		$completed_lists   = get_option( self::COMPLETED_OPTION, array() );
 		$completed_lists[] = $this->get_list_id();
-		update_option( self::COMPLETED_OPTION, $completed_lists );
+		update_option( self::COMPLETED_OPTION, $completed_lists, true );
```
Enable `autoload` for the option since we're using it on every WC admin page load.

chihsuan committed:
Fix tests

chihsuan committed:
Fix error

chihsuan committed:
Refactor task list state checks to improve visibility logic

- Replace `isTaskListCompletedOrHidden` with new `isTaskListActive` function
- Update multiple components to use the new task list state check
- Simplify conditional rendering across homescreen, header, and activity panel
- Modify deprecated tasks plugin rendering logic

chihsuan committed:
Add setupTaskListActive state to task list state management

- Introduce new `setupTaskListActive` property in task list state hooks
- Update default state and task list state retrieval to include active state
- Modify homescreen layout to use new task list state hook with active flag
- Enhance test coverage for new task list state property

chihsuan committed:
Update homescreen layout to use new task list state hook

- Add `setupTaskListHidden` to task list state hook
- Update task list rendering condition to use `setupTaskListHidden`
- Improve comment explaining task list state management

ilyasfoo reviewed (approved):
Nice work, @chihsuan! Tests well and LGTM.

Not caused by the PR, tasks API is being called again when opening up a task:

https://github.com/user-attachments/assets/2064da1e-8dad-4830-80c9-883b675afd85

chihsuan committed:
Simplify task screen detection logic in homescreen layout

- Refactor `isTaskScreen` condition to be more concise
- Remove redundant condition checking for task list presence
- Clarify comment explaining task screen detection

chihsuan committed:
Merge branch 'trunk' into perf/tasklist-api

rjchow commented on plugins/woocommerce/src/Admin/Features/OnboardingTasks/TaskList.php:
```diff
@@ -168,10 +168,23 @@ public function is_hidden() {
 	 */
 	public function is_visible() {
-		if ( ! $this->visible || $this->is_hidden() || ! count( $this->get_viewable_tasks() ) > 0 ) {
+		// If the task list is explicitly set to not be visible, return false.
```
feels like there may be a possibility of confusion on the difference between visible and hidden, to my understanding visible = dev intention to make an invisible task list, and hidden = user intent to dismiss the task list

chihsuan replied:
That's what I understand, too.

rjchow reviewed (approved):
Code looks good! I didn't have time to do a full test
